### PR TITLE
Add validator to pubsub and stop gossiping transactions twice

### LINKED ISSUE #93: Invalid transactions are gossiped in P2P network
Currently, gossiping relies purely on libp2p, with no validation. Even invalid transactions are propagated. Malicious node can spam the network with invalid transactions.
Ideally, this behavior should be limited; each node should not propagate invalid transaction to it's peers.

This can be implemented using [RegisterTopicValidator](https://pkg.go.dev/github.com/libp2p/go-libp2p-pubsub?utm_source=godoc#PubSub.RegisterTopicValidator). `TopicValidator` should call `CheckTx` to ensure, that only valid transactions are propagated.

### PR DESCRIPTION
**Pull request by evan-forbes**
PR to register a validator to gossip sub that embeds the mempools CheckTx method into gossip sub, along with stopping the node from rebroadcasting the tx after it is inserted into the mempool.

closes #93 and #94

### CODE REVIEW

File: node/node.go [hunk 1]
```diff
@@ -236,3 +193,19 @@ func (n *Node) EventBus() *lltypes.EventBus {
...
 	return n.proxyApp
 }
+
+func newProxyAppValidator(app proxy.AppConns) pubsub.Validator {
+	return func(c context.Context, i peer.ID, m *pubsub.Message) bool {
+		req := abci.RequestCheckTx{
+			Tx:   m.Data,
+			Type: abci.CheckTxType_New,
+		}
+		resp, err := app.Mempool().CheckTxSync(c, req)
```
Comment (evan-forbes) [inline node/node.go:203]: at least in BaseApp, calling CheckTx uses the same state until calling Commit resets it. So, I think calling CheckTx additional times could actually lead to some weird edge cases. Still not sure, but it's something I'm investigating.

File: rpcclient/local.go [hunk 2]
```diff
@@ -180,6 +190,13 @@ func (l *Local) BroadcastTxSync(ctx context.Context, tx types.Tx) (*ctypes.Resul
 	}
 	res := <-resCh
 	r := res.GetCheckTx()
+
+	// broadcast tx
+	err = l.node.P2P.GossipTx(ctx, tx)
+	if err != nil {
+		return nil, fmt.Errorf("tx added to local mempool but failed to gossip: %w", err)
+	}
```
Comment (evan-forbes) [inline rpcclient/local.go:198]: Wasn't 100% sure what to do in this situation. The tx could still be included in a block but an error will be returned as it's not gossiped.

wdyt?

File: node/node.go [hunk 3]
```diff
@@ -236,3 +193,19 @@ func (n *Node) EventBus() *lltypes.EventBus {
 func (n *Node) ProxyApp() proxy.AppConns {
 	return n.proxyApp
 }
+
+func newProxyAppValidator(app proxy.AppConns) pubsub.Validator {
```
Comment (tzdybal) [inline node/node.go:197]: I think `newTxValidator` would be better name, as this is validator for Tx's only.

**Review by tzdybal — CHANGES_REQUESTED** [review]
Thanks for your submission @evan-forbes! This looks great. I left some small comments.

File: node/node.go (hunk 3 above)
Comment (liamsi) [reply node/node.go:197]: Evan probably chose the name to highlight that this validator talks back to the proxy app. Depends which information is more relevant here (and if the resulting validator is put into a properly named var as well).

File: node/node.go (hunk 1 above)
Comment (liamsi) [reply node/node.go:203]: I think Evan is more worried about any side-effects _in the app_ of calling checkTx multiple times. @tzdybal are you saying that this won't happen anyways bc it gets prevented by the mempool that checkTx gets called multiple times effectively?

**Review by liamsi — COMMENTED** [review]
Thanks for taking this over Evan!

File: node/node.go (hunk 1 above)
Comment (evan-forbes) [reply node/node.go:203]: This PR is currently directly using the proxy app, which doesn't check to see if the tx is in the mempool. @liamsi  is correct, that's what I was worried about.

 As @tzdybal hinted, I think changing the validator to not directly use the app, and instead use the mempool's wrapper around `CheckTx`, is a significant improvement over what is already implemented in this PR. There is no reason to run `CheckTx` on a tx that we've already checked and is submitted to our local mempool, and using the mempool's wrapper around `CheckTx` fixes this.

File: rpcclient/local.go (hunk 2 above)
Comment (liamsi) [reply rpcclient/local.go:198]: Agree with Tomasz but what was the previous behaviour here (and what happens in tendermint in this case btw)?

File: rpcclient/local.go [hunk 4]
```diff
@@ -180,6 +190,13 @@ func (l *Local) BroadcastTxSync(ctx context.Context, tx types.Tx) (*ctypes.Resul
 	}
 	res := <-resCh
 	r := res.GetCheckTx()
+
+	// broadcast tx
```
[the remark this answers is not in the document]
Comment (evan-forbes) [reply rpcclient/local.go]: thanks for catching this! resolved in https://github.com/celestiaorg/optimint/pull/97/commits/533add6769ea3482266382b0ca52b39f7c9753d7

File: node/node.go [hunk 5]
```diff
@@ -236,3 +193,19 @@ func (n *Node) EventBus() *lltypes.EventBus {
...
+	return func(c context.Context, i peer.ID, m *pubsub.Message) bool {
```
[the remark this answers is not in the document]
Comment (evan-forbes) [reply node/node.go]: whoops, resolved in https://github.com/celestiaorg/optimint/pull/97/commits/d88c5fa8fc18fceba0440f93fb1e1b1d2e777bae

[the remark this answers is not in the document]
Comment (evan-forbes) [reply node/node.go]: resolved in https://github.com/celestiaorg/optimint/pull/97/commits/d88c5fa8fc18fceba0440f93fb1e1b1d2e777bae

File: mempool/mempool.go [hunk 6]
```diff
@@ -81,6 +81,8 @@ type Mempool interface {
 	// CloseWAL closes and discards the underlying WAL file.
 	// Any further writes will not be relayed to disk.
 	CloseWAL()
+
+	RemoveTxByKey(txKey [TxKeySize]byte, removeFromCache bool)
```
Comment (liamsi) [inline mempool/mempool.go:85]: Please add a comment.

**Review by evan-forbes — COMMENTED** [review]
this PR has been refactored so that the txValidator uses the mempool's wrapper around the `CheckTx` method instead of the proxyapps. 

It also stores the txValidator in the `Gossip` struct. While the solution in this PR works, I think it's not optimal and overly difficult to grok. In order to make easier to grok, it would require a rewrite of how `Gossip` structs are managed, as the validators need to be registered at the `pubsub.Pubsub` struct, which is not actually stored anywhere.

File: mempool/errors.go [hunk 7]
```diff
@@ -1,14 +1,15 @@
...
-var (
-	// ErrTxInCache is returned to the client if we saw tx earlier
-	ErrTxInCache = errors.New("tx already exists in cache")
-)
+// ErrTxInCache is returned to the client if we saw tx earlier
+type ErrTxInCache struct{}
+
+func (e ErrTxInCache) Error() string {
+	return "tx already exists in cache"
+}
```
[the remark this answers is not in the document]
Comment (liamsi) [reply mempool/errors.go:12]: For comparing errors, we should be using `errors.Is` which does not require a struct: https://pkg.go.dev/errors@master#Is

File: mempool/mempool.go (hunk 6 above)
Comment (evan-forbes) [reply mempool/mempool.go:87]: https://github.com/celestiaorg/optimint/pull/97/commits/ac95d91703b3673093160880aab643ff51e363b1

File: mempool/errors.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (evan-forbes) [reply mempool/errors.go]: good idea https://github.com/celestiaorg/optimint/pull/97/commits/0b62bb48b43ad146145794dffd90f37fc4350283

File: node/node.go [hunk 8]
```diff
@@ -187,14 +188,14 @@ func newTxValidator(pool mempool.Mempool, poolIDs *mempoolIDs, logger log.Logger
...
+		case errors.Is(err, mempool.ErrTxInCache):
 			return true
-		case mempool.ErrMempoolIsFull:
+		case errors.Is(err, mempool.ErrMempoolIsFull{}):
 			return true
-		case mempool.ErrTxTooLarge:
+		case errors.Is(err, mempool.ErrTxTooLarge{}):
 			return false
-		case mempool.ErrPreCheck:
+		case errors.Is(err, mempool.ErrPreCheck{}):
```
Comment (liamsi) [inline node/node.go]: We should make the errors more consistent here IMO. But let's do that in a separate PR. All these do not need to be structs but can instead be annotated using `fmt`: https://pkg.go.dev/fmt#hdr-Format_errors

**Comment by liamsi** [comment]
> While the solution in this PR works, I think it's not optimal and overly difficult to grok. In order to make easier to grok, it would require a rewrite of how Gossip structs are managed, but that should probably be completed in a separate PR after more discussion.

If we merge this PR as is, it would be good to open issue explaining how this would look like.

File: rpcclient/local.go (hunk 2 above)
Comment (liamsi) [reply rpcclient/local.go]: I think it is wrong to remove a valid Tx from the mempool only bc gossiping failed. What if there is a temporary networking error?
https://github.com/celestiaorg/optimint/blob/4097620751907e1bf0f1a4b5839d34b7ffcd5680/rpcclient/local.go#L194-L197

Comment (liamsi) [reply rpcclient/local.go]: It's really hard to see that the rest of the code is correct IMO.
Also, why does the rpcclient do the broadcasting here?

**Review by liamsi — DISMISSED** [review]
It's hard to see that this actually works as intended. Also, is it intentional that the local rpc client handles p2p gossiping? Also, removing a tx from the mempool immediately bc it failed to gossip does not seem right imo.

File: node/node.go [hunk 9]
```diff
@@ -119,69 +125,6 @@ func NewNode(ctx context.Context, conf config.NodeConfig, nodeKey crypto.PrivKey
 	return node, nil
 }
 
-func (n *Node) mempoolReadLoop(ctx context.Context) {
-	for {
-		select {
-		case tx := <-n.incomingTxCh:
```
Comment (liamsi) [inline node/node.go:125]: Is the `incomingTxCh` unused now?

File: rpcclient/local.go [hunk 10]
```diff
@@ -180,6 +190,15 @@ func (l *Local) BroadcastTxSync(ctx context.Context, tx types.Tx) (*ctypes.Resul
 	}
 	res := <-resCh
 	r := res.GetCheckTx()
+
+	if r.Code == abci.CodeTypeOK {
+		err = l.node.P2P.GossipTx(ctx, tx)
+		if err != nil {
+			l.node.Mempool.RemoveTxByKey(mempool.TxKey(tx), true)
```
[the remark this answers is not in the document]
Comment (liamsi) [reply rpcclient/local.go:197]: Oh, I see. This now means that the app said this Tx is valid but no-one will ever see it because gossiping failed (and we out self did not receive it via gossiping but locally)? Is that right?

File: rpcclient/local.go (hunk 2 above)
Comment (liamsi) [reply rpcclient/local.go]: Oh, I think I see: https://github.com/celestiaorg/optimint/pull/97#discussion_r708244572
We should document this better IMO.

File: node/node.go (hunk 9 above)
Comment (evan-forbes) [reply node/node.go:125]: that is a great point, currently yes it is. 

@tzdybal do you think we can remove that, or do we have other plans for it? What about `GossipHandler`s as well, do we still need those?

Perhaps we can address those in the future proposals that address #100 ?

File: rpcclient/local.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (evan-forbes) [reply rpcclient/local.go:197]: yeah, that's correct

File: rpcclient/local.go (hunk 2 above)
Comment (evan-forbes) [reply rpcclient/local.go]: > what happens in tendermint in this case btw)?

afaict, this scenario doesn't occur in tendermint, as the tx is gossiped as part of the mempool reactor, so as long as the tx is in the mempool, it is being gossiped. 

This is not the case for here. If the user submits a tx and if fails to be gossiped, then it will remain in the local mempool and will never be gossiped. In fact, unless we remove it from the mempool, the user will be unable to try and gossip that transaction again using the same node, because when they do `CheckTx` will return an error indicating that the tx already exists in the mempool.

File: rpcclient/local.go [hunk 11]
```diff
@@ -180,6 +190,22 @@ func (l *Local) BroadcastTxSync(ctx context.Context, tx types.Tx) (*ctypes.Resul
...
+	// Note: we have to do this here because, unlike the tendermint mempool reactor, there
+	// is no routine that gossips transactions after they enter the pool
+	if r.Code == abci.CodeTypeOK {
+		err = l.node.P2P.GossipTx(ctx, tx)
+		if err != nil {
+			// the transaction must be removed from the mempool if it cannot be gossiped.
+			// if this does not occur, then the user will not be able to try again using
+			// this node, as the CheckTx call above will return an error indicating that
+			// the tx is already in the mempool
+			l.node.Mempool.RemoveTxByKey(mempool.TxKey(tx), true)
```
Comment (evan-forbes) [inline rpcclient/local.go:204]: added some docs

File: rpcclient/local_test.go [hunk 12]
```diff
@@ -225,11 +246,15 @@ func TestMempool2Nodes(t *testing.T) {
...
+	assert.NotNil(resp)
+	// broadcast the bad Tx, this should not be propogated or added to the local mempool
+	resp, err = local.BroadcastTxSync(ctx, []byte("good"))
 	assert.NoError(err)
 	assert.NotNil(resp)
-
 	time.Sleep(1 * time.Second)
 
-	assert.Equal(node2.Mempool.TxsBytes(), int64(len("foobar")))
+	assert.Equal(node2.Mempool.TxsBytes(), int64(len("good")))
```
Comment (evan-forbes) [inline rpcclient/local_test.go:259]: modified the test that checks that the mempool is working to also check that invalid transactions do not get gossiped.

**Review by evan-forbes — COMMENTED** [review]
made some more changes to incorporate more feedback :pray: :slightly_smiling_face:

File: node/node.go (hunk 9 above)
Comment (tzdybal) [reply node/node.go:125]: My comment was discarded somehow O_o

I think it's reasonable, to keep handler and validator separate (see https://github.com/celestiaorg/optimint/issues/100#issuecomment-921679366).

File: go.mod [hunk 13]
```diff
@@ -6,7 +6,6 @@ require (
 	github.com/celestiaorg/celestia-app v0.0.0-00010101000000-000000000000
 	github.com/cosmos/cosmos-sdk v0.40.0-rc5
 	github.com/dgraph-io/badger/v3 v3.2011.1
-	github.com/fortytw2/leaktest v1.3.0
```
Comment (liamsi) [inline go.mod:9]: Thanks for tidying up go.mod. Looks like this wasn't use anywhere anyways.

File: node/mempool.go [hunk 14]
```diff
@@ -61,8 +61,8 @@ func (ids *mempoolIDs) Reclaim(peer peer.ID) {
 
 // GetForPeer returns an ID for the peer. ID is generated if required.
 func (ids *mempoolIDs) GetForPeer(peer peer.ID) uint16 {
-	ids.mtx.RLock()
-	defer ids.mtx.RUnlock()
+	ids.mtx.Lock()
+	defer ids.mtx.Unlock()
```
Comment (liamsi) [inline node/mempool.go:65]: Nice find. It's obviously not just read here.

File: node/node.go (hunk 9 above)
Comment (liamsi) [reply node/node.go:125]: regarding the initial question: `incomingTxCh`is still unused now. @tzdybal are you advocating to keep it because keeping the "handler and validator separate" would reintroduce that channel?

File: node/node.go [hunk 15]
```diff
@@ -235,3 +174,33 @@ func (n *Node) EventBus() *lltypes.EventBus {
...
+			SenderP2PID: corep2p.ID(m.GetFrom()),
+		})
+		switch {
+		case errors.Is(err, mempool.ErrTxInCache):
+			return true
+		case errors.Is(err, mempool.ErrMempoolIsFull{}):
+			return true
+		case errors.Is(err, mempool.ErrTxTooLarge{}):
+			return false
+		case errors.Is(err, mempool.ErrPreCheck{}):
```
Comment (liamsi) [inline node/node.go:197]: Orthogonal to this PR but the errors should consistently be vars and wrapped using `fmt.Errorf` and `%w` to decorate them with the necessary info.

File: rpcclient/local.go [hunk 16]
```diff
@@ -120,6 +120,12 @@ func (l *Local) BroadcastTxCommit(ctx context.Context, tx types.Tx) (*ctypes.Res
 		}, nil
 	}
 
+	// broadcast tx
+	err = l.node.P2P.GossipTx(ctx, tx)
+	if err != nil {
+		return nil, fmt.Errorf("tx added to local mempool but failure to broadcast: %w", err)
+	}
+
```
Comment (liamsi) [inline rpcclient/local.go:128]: I still think it's weird that the p2p gossiping is now triggered in rpcclient but yeah let's fix this in a followup PR.

**Review by liamsi — COMMENTED** [review]
OK I think with #100 this can later be improved and be merged as is. I'll not yet approve this and let @mattdf give this another pass.

**Review by tzdybal — APPROVED** [review]
This is much better than existing code, but it's not the end of the journey (see #100).

**Review by liamsi — APPROVED** [review]
